### PR TITLE
Debugging - DX to help make package devs and others know what contexts can be consumed.

### PR DESCRIPTION
**Pull request by warrenbuckley**
Hacking the planet - trying to add some DEBUG pane to help display what Contexts are available or events are firing..

## Notes
The `<umb-debug></umb-debug>` component has been put inside the dashboard of Published Status in the Settings section which adds a button in its place and toggles a DIV containing the list of Context aliases its collected when crawling up through the tree until it gets to the root node of `<umb-app>`

The idea is that implementors or package developers can place this component inside something they are working on such as dashboard, property editor or workspace to understand and know what contexts are available to them to use.

## First Rough UI
image

image

## What next ?
- [x] Could the list of Context's display the useable methods/functions to help with discoverability
- [x] Add support for showing in modal
- [ ] Add Storybook story to showcase how it works
- [ ] (Optional) Add to README

### CODE REVIEW

**Review by iOvergaard — APPROVED** [review]
LGTM

File: src/backoffice/shared/components/debug/debug.modal.element.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+			ul {
+				margin-top: 0;
+			}
+		`,
+	];
+
+
+	// the modal handler will be injected into the element when the modal is opened.
+	@property({ attribute: false })
+	modalHandler?: UmbModalHandler;
```
Comment (iOvergaard) [inline src/backoffice/shared/components/debug/debug.modal.element.ts:60]: I think it gets this property automatically from its base element

File: src/backoffice/shared/components/debug/debug.modal.element.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
+import { css, html, nothing, TemplateResult } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { UUITextStyles } from '@umbraco-ui/uui-css';
+import { UmbModalHandler, UmbModalLayoutElement } from '@umbraco-cms/modal';
+
+@customElement('umb-debug-modal-layout')
+export class UmbDebugModalLayout extends UmbModalLayoutElement {
```
Comment (iOvergaard) [inline src/backoffice/shared/components/debug/debug.modal.element.ts:7]: We need to send in a generic here to imply what the data might be

```suggestion
export class UmbDebugModalLayout extends UmbModalLayoutElement<any> {
```

would do for now

File: src/backoffice/shared/components/debug/debug.element.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+				margin: 0;
+			}
+		`,
+	];
+
+	@property({ reflect: true, type: Boolean })
+	enabled = false;
+
+	@property({ reflect: true, type: Boolean })
+	useDialog = false;
```
Comment (iOvergaard) [inline src/backoffice/shared/components/debug/debug.element.ts:59]: Let's call this "dialog" so it resembles the wording of the "enabled" property and looks better when using it

```suggestion
	dialog = false;
```

File: src/backoffice/shared/components/debug/debug.element.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+				// of the this event tha has been raised to assign the contexts
+				// back to this property of the WebComponent
+				this.contexts = contexts;
+			})
+		);
+	}
+
+	render() {
+		if (this.enabled) {			
+			return this.useDialog ? this._renderDialog() : this._renderPanel();
```
Comment (iOvergaard) [inline src/backoffice/shared/components/debug/debug.element.ts:95]:
```suggestion
			return this.dialog ? this._renderDialog() : this._renderPanel();			
```

File: src/backoffice/settings/dashboards/published-status/dashboard-published-status.element.ts [hunk 5]
```diff
@@ -133,6 +133,7 @@ export class UmbDashboardPublishedStatusElement extends UmbLitElement {
 
 	render() {
 		return html`
+			<umb-debug enabled useDialog></umb-debug>
```
Comment (iOvergaard) [inline src/backoffice/settings/dashboards/published-status/dashboard-published-status.element.ts:136]:
```suggestion
			<umb-debug enabled dialog></umb-debug>
```

File: src/backoffice/settings/dashboards/examine-management/views/modal-views/fields-viewer.element.ts [hunk 6]
```diff
@@ -16,11 +16,7 @@ export class UmbModalLayoutFieldsViewerElement extends UmbModalLayoutElement<Sea
 				display: flex;
 				flex-direction: column;
 				height: 100%;
-				background-color: var(--uui-color-surface);
```
Comment (iOvergaard) [inline src/backoffice/settings/dashboards/examine-management/views/modal-views/fields-viewer.element.ts:19]: What does this change do?

File: src/backoffice/shared/components/debug/debug.modal.element.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+			return nothing;
+		}
+
+		const aliases = [];
+		for (const [alias, instance] of this.data.contexts) {
+			aliases.push(
+				html`
+				<div class="context">
+					<h3>${alias} <span>${typeof instance}</span></h3>
+					${this._renderInstance(instance)}
```
Comment (iOvergaard) [inline src/backoffice/shared/components/debug/debug.modal.element.ts:94]: Consider letting this modal layout render `this.data?.content` and reuse the same templates from the outer DebugElement class instead. This way you avoid duplicating the logic. See ModalLayoutConfirmElement for inspiration.

Alternatively, consider putting the rending into a new element (UmbDebugRenderElement) and use that for both inline and modal modes.

```
<slot></slot>
```

File: src/backoffice/shared/components/index.ts [hunk 8]
```diff
@@ -30,3 +30,6 @@ import './input-document-picker/input-document-picker.element';
 
 import './empty-state/empty-state.element';
 import './color-picker/color-picker.element';
+
+import './debug/debug.element';
+import './debug/debug.modal.element';
```
Comment (iOvergaard) [inline src/backoffice/shared/components/index.ts:35]: Consider importing the `debug.modal.element.ts` lazily inside `debug.element.ts` only when the component has to run in modal mode.

Comment (warrenbuckley) [reply src/backoffice/shared/components/index.ts]: @iOvergaard I have moved it into the main `debug.element` but I am unsure how to lazy load it and couldn't find an example in the codebase. Can you give me a pointer please.

File: src/backoffice/settings/dashboards/examine-management/views/modal-views/fields-viewer.element.ts (hunk 6 above)
Comment (warrenbuckley) [reply src/backoffice/settings/dashboards/examine-management/views/modal-views/fields-viewer.element.ts:19]: Well --uui-color-surface is white, but there was no need to explicitly set the background color to be white, when the parent is white.

File: src/backoffice/shared/components/debug/debug.modal.element.ts (hunk 7 above)
Comment (warrenbuckley) [reply src/backoffice/shared/components/debug/debug.modal.element.ts]: I have done as you suggested and made it more generic and now pass the `TemplateResult` as a property inside the data object of the dialog/modal.

**Comment by warrenbuckley** [comment]
@iOvergaard the last few things for me that I need some pointers with are
* Storybook stories
* Advice on how to lazyload the import of the dialog

## Possible Improvements
* As the modal layout markup is passed to it - could a very basic generic modal layout be created for people to create basic modals that are used for display only ?

**Review by iOvergaard — APPROVED** [review]
Fixed the modal so it loads on demand now. LGTM!

**Pull request merged**